src/Storage/Device/Local.php

```diff
@@ -69,24 +71,68 @@ public function getPath($filename): string
      *
      * @param string $source
      * @param string $path
+     * @param int $chunk
+     * @param int $chunks
+     * @param array $metadata
      *
      * @throws \Exception
      *
-     * @return bool
+     * @return int
      */
-    public function upload($source, $path): bool
+    public function upload($source, $path, $chunk = 1, $chunks = 1, &$metadata = []): int
     {
         if (!\file_exists(\dirname($path))) { // Checks if directory path to file exists
             if (!@\mkdir(\dirname($path), 0755, true)) {
                 throw new Exception('Can\'t create directory: ' . \dirname($path));
             }
         }
 
-        if (\move_uploaded_file($source, $path)) {
-            return true;
+        //move_uploaded_file() verifies the file is not tampered with
+        if($chunks == 1) {
+            if (!\move_uploaded_file($source, $path)) {
+                throw new Exception('Can\'t upload file ' . $path);
+            }
+            return $chunks;
         }
+        $tmp = \dirname($path) . "/tmp/chunks.log";
 
-        return false;
+        if (!\file_exists(\dirname($tmp))) { // Checks if directory path to file exists
+            if (!@\mkdir(\dirname($tmp), 0755, true)) {
+                throw new Exception('Can\'t create directory: ' . \dirname($tmp));
+            }
+        }
+        if(!file_put_contents($tmp, "{$chunk}\n", FILE_APPEND)) {
+            throw new Exception('Can\'t write chunk log ' . $tmp);
+        }
+
+        $chunkLogs = file($tmp);
+        if(!$chunkLogs) {
+            throw new Exception("Unable to read chunk log " . $tmp);
+        }
+
+        $chunksReceived = count(file($tmp));
+
+        if(!\rename($source, dirname($tmp) . DIRECTORY_SEPARATOR . pathinfo($path, PATHINFO_FILENAME) . ".part.{$chunk}")) {
+            throw new Exception('Failed to write chunk ' . $chunk);
+        }
+        
+        if ($chunks == $chunksReceived) {
+            for($i = 0; $i < $chunks; $i++) {
+                $part = dirname($tmp) . DIRECTORY_SEPARATOR . pathinfo($path, PATHINFO_FILENAME) . ".part.{$i}";
+                $data = file_get_contents($part);
+                if(!$data) {
+                    throw new Exception("Failed to read chunk " . $part);
+                }
+
+                if(!file_put_contents($path, $data, FILE_APPEND)) {
+                    throw new Exception('Failed to append chunk ' . $part);
+                }
+                \unlink($part);
```

Comment:
Are these `unlink()` calls necessary? If so, should `$data` be unlinked?

Reply:
$data is just a string. `unlink()` I believe is used for files. 

I think you're talking about `unset()`

Reply:
@kodumbeats yes unlink() is necessary to delete the part file and log file once the upload is complete